src/editor/nodes/ParticleNode.js

```diff
@@ -0,0 +1,416 @@
+    this.lifetimes = lifetimes;
+    this.colors = colors;
+
+    //console.log(this);
+  }
+
+  async load(src) {
+    const nextSrc = src || "";
+    if (nextSrc === this._canonicalUrl) {
+      return;
+    }
+
+    this._canonicalUrl = nextSrc;
+
+    const { accessibleUrl } = await this.editor.api.resolveMedia(src);
+    console.log("accessibleUrl: " + accessibleUrl);
+    this.material.uniforms.texture.value = await new Promise((resolve, reject) => {
+      new THREE.TextureLoader().load(accessibleUrl, resolve, null, e =>
+        reject(`Error loading Image. ${eventToMessage(e)}`)
+      );
+    });
+    this.material.uniforms.texture.value.flipY = false;
+
+    return this;
+  }
+
+  onUpdate(dt) {
+    const position = this.geometry.attributes.position.array;
+    const color = this.geometry.attributes.color.array;
+    const customAngle = this.geometry.attributes.customAngle.array;
+    const colorFactor = [] || 0;
+    let angleVel = 0;
+
+    const velFactor = [] || 1;
+
+    for (let i = 0; i < this.particleCount; i++) {
+      this.ages[i] += dt;
+
+      velFactor[i] = this.clamp(0, 1, this.ages[i] / this.lifetimes[i]);
+      colorFactor[i] = this.clamp(0, 1, this.ages[i] / this.lifetimes[i]);
+      position[i * 4 + 3] = this.size;
+
+      if (this.ages[i] < 0) {
+        color[i * 4] = this.startColor.r;
+        color[i * 4 + 1] = this.startColor.g;
+        color[i * 4 + 2] = this.startColor.b;
+        color[i * 4 + 3] = this.startColor;
+        velFactor[i] = 0;
+        continue;
+      }
+
+      if (this.ages[i] > this.lifetimes[i]) {
+        position[i * 4] = this.initialPositions[i * 3];
+        position[i * 4 + 1] = this.initialPositions[i * 3 + 1];
+        position[i * 4 + 2] = -2; //this.initialPositions[i * 3 + 2];
+        velFactor[i] = 0;
+        this.ages[i] = this.initialAges[i];
+        colorFactor[i] = 0;
+
+        continue;
+      }
+
+      switch (this.velocityCurve) {
+        case "Linear":
+          break;
+        case "Ease-in":
+          velFactor[i] = this.EaseIn(velFactor[i]);
+          break;
+        case "Ease-out":
+          velFactor[i] = this.EaseOut(velFactor[i]);
+          break;
+        case "Ease-in, out":
+          velFactor[i] = this.EaseInOut(velFactor[i]);
+          break;
+      }
+
+      this.velocities[i * 3] = this.lerp(this.velocity.x, this.endVelocity.x, velFactor[i]);
+      this.velocities[i * 3 + 1] = this.lerp(this.velocity.y, this.endVelocity.y, velFactor[i]);
+      this.velocities[i * 3 + 2] = this.lerp(this.velocity.z, this.endVelocity.z, velFactor[i]);
+
+      position[i * 4] += this.velocities[i * 3] * dt;
+      position[i * 4 + 1] += this.velocities[i * 3 + 1] * dt;
+      position[i * 4 + 2] += this.velocities[i * 3 + 2] * dt;
+
+      angleVel = this.angularVelocity;
+      customAngle[i] += angleVel * 0.01745329251 * dt;
+
+      switch (this.colorCurve) {
+        case "Even":
+          colorFactor[i] = this.Even(colorFactor[i]);
```

Comment:
You can probably maintain a map of these different curve types like:
```
const Curves = {
  Even() {}
  EaseIn() {}
  ...
};
```

And then look up the function like this:
```
const colorCurveFn = Curves[this.colorCurve];
colorFactor[i] = colorCurveFn(colorFactor[i]);
```